ipv8/REST/dht_endpoint.py

```diff
@@ -100,53 +107,67 @@ def reconstruct_all_blocks(self, block_chunks, public_key):
 
-                # If we reached this point, it means the latest_block is novel
-                my_private_key = self.trustchain.my_peer.key
+                signature = my_private_key.signature(pre_signed_content)
+                blob_chunk = self.serializer.pack_multiple(
+                    DHTBlockPayload(signature, self.block_version, chunk_idx, total_chunks, chunk).to_pack_list())
 
-                # Get the total number of chunks in this blocks
-                total_blocks = len(latest_block) // self.CHUNK_SIZE
-                total_blocks += 1 if len(latest_block) % self.CHUNK_SIZE != 0 else 0
+                # Try to add the current chunk to the DHT; if it works, move to the next, otherwise retry
+                d = self.dht.store_value(self._hashed_dht_key, blob_chunk[0])
+                d.addCallback(publish_chunk, slice_pointer + self.CHUNK_SIZE, chunk_idx + 1, 1)
+                d.addErrback(publish_chunk, slice_pointer, chunk_idx, chunk_attempt + 1)
 
-                # To make this faster we'll use addition instead of multiplication, and use a pointer
-                slice_pointer = 0
+            def on_failure_find_values(_):
+                # If for any reason, finding the values under a key failed, retry this entire operation
+                self.publish_latest_block(block, attempt + 1)
 
-                for i in range(total_blocks):
-                    chunk = latest_block[slice_pointer: slice_pointer + self.CHUNK_SIZE]
-                    slice_pointer += self.CHUNK_SIZE
-                    signature = my_private_key.signature(
-                        str(self.block_version).encode('utf-8') + str(i).encode('utf-8')
-                        + str(total_blocks).encode('utf-8') + chunk)
-                    blob_chunk = self.serializer.pack_multiple(
-                        DHTBlockPayload(signature, self.block_version, i, total_blocks, chunk).to_pack_list())
+            def on_success_find_values(block_chunks):
+                new_blocks, _ = self.reconstruct_all_blocks([x[0] for x in block_chunks],
+                                                            self.trustchain.my_peer.public_key)
 
-                    self.dht.store_value(self._hashed_dht_key, blob_chunk[0]).addErrback(on_failure)
+                # Check if the block we want to publish is actually a duplicate
+                if latest_block in new_blocks.values():
+                    return
 
-                self.block_version += 1
+                # We now know the block is novel, so we'll try to publish it
+                deferLater(reactor, 0, publish_chunk, None, 0, 0)
 
-            self.dht.find_values(self._hashed_dht_key).addCallbacks(on_success, on_failure)
+            # On the off chance that the block is actually empty, avoid publishing it
+            if total_chunks > 0:
+                self.dht.find_values(self._hashed_dht_key).addCallbacks(on_success_find_values, on_failure_find_values)
```

Comment:
Quick question: why are we first doing a `find_values` before storing stuff to the DHT? Isn't the `BlockListener` called only once for every block? If so, I'm not sure whether this is needed. Other than that, this PR looks ready to merge.

Reply:
It's just a bit of defensive programming. Indeed, I don't think there's a significant risk of a block being duplicated (since a callback is called only once for a new block), but in an effort to avoid trying to republish the same block twice (if somehow that might happen), I added that part in as well. I'll remove it, since it's likely not very useful.

Reply:
Should be fixed now.